src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,77 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, vect, stride, offset)
+    for i in offset:stride:length(vect)
+        push!(col, vect[i])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    allunique(indexcols) || throw(ArgumentError("Elements of $indexcols must be unique"))
+    sum(DataFrames.nonunique(df[:,indexcols]))>0 && @warn "duplicate rows in input; expand will only return unique combinations"
+    colind = index(df)[indexcols]
+
+    dummydf = similar(select(df, indexcols, copycols=false),0)
+
+    # Create a vect of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        if sum(ismissing.(df[col])) != 0
+            tempcol = vcat(levels(df[col]), missing)
+        else
+            tempcol = levels(df[col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(flatten(product(values(uniqueVals)...)))
+    dfsize = div(length(collected),length(indexcols))
+
+    for i in 1:length(indexcols)
+        _expandhelper(dummydf[!,i], collected, length(indexcols), i)
+    end
+
+    return dummydf
+end
+
+function complete(df::AbstractDataFrame, indexcols::Array{Symbol,1}; fill=missing::Any, replaceallmissing=false)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    allunique(indexcols) || throw(ArgumentError("Elements of $indexcols must be unique"))
+    sum(DataFrames.nonunique(df[:,indexcols]))>0 && throw(ArgumentError("duplicate rows in input"))
+    colind = index(df)[indexcols]
+
+    # Expand the input df and left join
+    expanded = expand(df, indexcols)
+    expanded = join(expanded, df; on = indexcols, kind = :left, indicator = :source)
+
+    # Replace missing values with the fill
+    if !ismissing(fill)
+        for n in names(expanded)
+            if (n ∉ indexcols) && (n != :source)
```

Comment:
the `:source` test is incorrect as the name might be duplicated. Better use the last column test (and then you can iterate not names but indices of a `DataFrame`).

Reply:
That's a good point! Would using `names(df)[end]` be an appropriate solution to this?

Reply:
Tested it with a df with a column `:source` and it appears to work.

Reply:
yes `end` is OK, as the indicator column is always added at the end.